### ISSUE TITLE
Ubuntu24.04 error: cannot load such file -- net/smtp (LoadError)

### ISSUE THREAD

**Issue by chhtz** [rock-core/autoproj#408]
Not sure if that is already on the agenda. I'm trying to bootstrap a project on Ubuntu24 using:

    wget http://rock-robotics.org/master/autoproj_bootstrap
    ruby autoproj_bootstrap git ${some bootstrap url}

Full output (HOME_DIR and BOOTSTRAP_DIR are valid directories of course):

```
Detected 'gem' to be /usr/bin/gem3.2
Detected bundler at /HOME_DIR/.local/share/autoproj/gems/ruby/3.2.0/bin/bundle
Installing autoproj in /HOME_DIR/.local/share/autoproj/gems/ruby/3.2.0
Fetching gem metadata from https://rubygems.org/.......
Resolving dependencies...
Bundle complete! 1 Gemfile dependency, 27 gems now installed.
Bundled gems are installed into `/HOME_DIR/.local/share/autoproj/gems`
starting the newly installed autoproj for stage2 install
<internal:/usr/lib/ruby/vendor_ruby/rubygems/core_ext/kernel_require.rb>:38:in `require': cannot load such file -- net/smtp (LoadError)
	from <internal:/usr/lib/ruby/vendor_ruby/rubygems/core_ext/kernel_require.rb>:38:in `require'
	from /HOME_DIR/.local/share/autoproj/gems/ruby/3.2.0/gems/autobuild-1.24.0/lib/autobuild.rb:14:in `<top (required)>'
	from <internal:/usr/lib/ruby/vendor_ruby/rubygems/core_ext/kernel_require.rb>:38:in `require'
	from <internal:/usr/lib/ruby/vendor_ruby/rubygems/core_ext/kernel_require.rb>:38:in `require'
	from /HOME_DIR/.local/share/autoproj/gems/ruby/3.2.0/gems/autoproj-2.17.0/lib/autoproj/ops/install.rb:802:in `stage2'
	from /HOME_DIR/.local/share/autoproj/gems/ruby/3.2.0/gems/autoproj-2.17.0/lib/autoproj/cli/main.rb:565:in `install_stage2'
	from /HOME_DIR/.local/share/autoproj/gems/ruby/3.2.0/gems/thor-1.3.2/lib/thor/command.rb:28:in `run'
	from /HOME_DIR/.local/share/autoproj/gems/ruby/3.2.0/gems/thor-1.3.2/lib/thor/invocation.rb:127:in `invoke_command'
	from /HOME_DIR/.local/share/autoproj/gems/ruby/3.2.0/gems/thor-1.3.2/lib/thor.rb:538:in `dispatch'
	from /HOME_DIR/.local/share/autoproj/gems/ruby/3.2.0/gems/thor-1.3.2/lib/thor/base.rb:584:in `start'
	from /HOME_DIR/.local/share/autoproj/gems/ruby/3.2.0/gems/autoproj-2.17.0/bin/autoproj:8:in `<top (required)>'
	from /BOOTSTRAP_DIR/.autoproj/bin/autoproj:49:in `load'
	from /BOOTSTRAP_DIR/.autoproj/bin/autoproj:49:in `<main>'
autoproj_bootstrap:807:in `call_stage2': failed to execute autoproj install-stage2 (RuntimeError)
	from autoproj_bootstrap:846:in `<main>'
```

Are there any known workarounds? I tried manually installing 

    gem install -i ~/.local/share/autoproj/gems/ net-smtp

Or adding that to autoproj's Gemfile without success. 

I guess I don't need any smtp-support, but probably this won't be the only issue.

**Comment by g-arjones** [comment]
Can you please try with https://github.com/rock-core/autoproj/raw/refs/heads/master/bin/autoproj_bootstrap instead?

**Comment by g-arjones** [comment]
Nevermind... There's a PR still pending on autobuild.

**Comment by chhtz** [comment]
Thanks for the fast response! 
It is indeed still the same error (except for some line numbers). I assume you mean this PR https://github.com/rock-core/autobuild/pull/127? 
Can I test locally with a fork/branch of autobuild?

**Comment by chhtz** [comment]
Nice that works! (I downloaded that and replaced the branch name of autobuild inside the script). And in fact bootstrapping succeeds, but `aup` fails at some point with:

```
[...]
Command failed
the 'binding_of_caller' osdep refers to another osdep, 'ruby-dev', which does not seem to exist: Autoproj::OSPackageResolver::InvalidRecursiveStatement
```
I could resolve this by copying:

    cp ./autoproj/remotes/orocos.toolchain/orocos.osdeps-ruby3{0,2}

After that, I think I only have dependency issues related to the actual bootstrap (e.g., not finding `libqwt5-qt4-dev` ...), i.e., from my side I'm fine with closing this once the related PR is merged.

**Comment by ndahn** [comment]
FYI, one major issue is that recent versions of ruby abandoned `File.exists?` in favor of `File.exist?` (no s)...

**Comment by yarohenning** [comment]
Hi, I'm trying autopoj_bootstrap-dev on 24.04 and it seems autoproj is no longer installed as gem. This leads to problems when using autoproj in ruby scripts. e.g. list_rock_osdeps.rb. (Varying steps 5,6, 7 doesn't influence the result)

Steps to reproduce:
1. ``docker run -ti ubuntu:24.04 bash`` (Following steps happen inside the docker container)
2. ``apt update && apt install -y git wget ruby-dev``
3. mkdir -p /opt/workspace && cd /opt/workspace
4. ``wget https://rock-robotics.org/autoproj_bootstrap-dev``
5. (tried with and without this step) Using the other autobuild branch
```bash
sed -i '271d' autoproj_bootstrap-dev
   sed -i '271i \                 "gem \\\"autobuild\\\", git: \\\"https://github.com/rock-core/autobuild.git\\\", branch: \\\"run_workflows_on_ruby3\\\"",' autoproj_bootstrap-dev
```
6. ``ruby autoproj_bootstrap-dev`` or ``ruby autoproj_bootstrap-dev git <buildconf>``
7. (tried with and without this step) `source env.sh`
8. Inspect output of ``gem list``: neither autobuild nor autoproj is listed there
9. Create a test file: `echo 'require "autoproj"' > test.rb` 
10. ``ruby test.rb`` will return```internal:/usr/lib/ruby/vendor_ruby/rubygems/core_ext/kernel_require.rb>:86:in `require': cannot load such file -- autoproj (LoadError)
        from <internal:/usr/lib/ruby/vendor_ruby/rubygems/core_ext/kernel_require.rb>:86:in `require'
        from test.rb:1:in `<main>'
```

**Comment by planthaber** [comment]
@yarohenning When you use `wget https://rock-robotics.org/autoproj_bootstrap-dev` the -dev set the source installation of autoproj/autobuild, when you want to use the gem, use `wget https://rock-robotics.org/autoproj_bootstrap` (without -dev)



I could bootstrap autoproj on 24.04 with the script from the autoproj git: (https://github.com/rock-core/autoproj/blob/master/bin/autoproj_bootstrap) by adding "gem "net-smtp" here:  https://github.com/rock-core/autoproj/blob/master/bin/autoproj_bootstrap#L267-L271

```ruby 
            def default_gemfile_contents(autoproj_version = ">= 2.17.0")
                ["source \"#{gem_source}\"",
                 "ruby \"#{RUBY_VERSION}\" if respond_to?(:ruby)",
                 "gem \"autoproj\", \"#{autoproj_version}\"",
                 "gem \"net-smtp\""].join("\n")
            end
```

At least that fixes the net/smtp error for the autoproj bootstrap with the non -dev script.

**Comment by chhtz** [comment]
@planthaber That's nice and actually much more convenient than the current `autoproj_boostrap-dev` workaround. Can we patch the official `autoproj-bootstrap` to include `net-smtp`? Or maybe on a branch of `autoproj` until the required autobuild changes are upstream?

**Comment by planthaber** [comment]
I guess way to go is finally merge https://github.com/rock-core/autobuild/pull/127. 

It looks to me that https://github.com/rock-core/autobuild/blob/run_workflows_on_ruby3/autobuild.gemspec#L27 will also fix the issue.

**Comment by chhtz** [comment]
This successfully bootstraps for me now:
```
wget http://rock-robotics.org/master/autoproj_bootstrap
sed -i '270i  "gem \\"net-smtp\\"",' autoproj_bootstrap 
ruby autoproj_bootstrap git {SOME_BOOTSTRAP_URL}
```

**Comment by chhtz** [comment]
> I guess way to go is finally merge [rock-core/autobuild#127](https://github.com/rock-core/autobuild/pull/127).
> 
> It looks to me that https://github.com/rock-core/autobuild/blob/run_workflows_on_ruby3/autobuild.gemspec#L27 will also fix the issue.

Yes, these are actually the same (but probably only that line in the gemspec will be easier to merge in an independent PR?)

**Comment by chhtz** [comment]
Fixed with autobuild 1.25 (I managed to do a bootstrap from without any manual patches on Ubuntu24 now).

Anyone who still experiences issues (with a fresh bootstrap) feel free to re-open and describe the problem (or open a new issue).
